### PR TITLE
Update TensorFlow pin (08-2022)

### PR DESCRIPTION
**Pull request by wonjoolee95**
[WIP] Update TensorFlow pin (08-2022)

Changes include:
- Updated path compiler/mlir/hlo/WORKSPACE to compiler/xla/mlir_hlo/WORKSPACE.
  - Reason: TF xla/hlo directory was moved (https://github.com/tensorflow/tensorflow/commit/6ed5af7784f3f1eaaea77cf1224b588e2521cf73). 
- Updated path tensorflow/core/platform/default to tensorflow/tsl/platform/default
  - Reason: TF //tensorflow/core/platform/default was moved to //tensorflow/tsl/platform/default as of https://github.com/tensorflow/tensorflow/commit/d727fc2b1ef78961b2eba7dbbfba0094ea2c8a0f
- Updated ConsumeValueOrDie() to value() in xla_client
  - Reason: `ConsumeValueOrDie()` was deprecated in favor of value() in TF as of https://github.com/tensorflow/tensorflow/commit/9dcc8d0cfa870090753a8108ef0d16ff1ebde049
- Update PjRtExecutable to PjRtLoadedExecutable
  - Reason: PjRtExecutable was introduced as a new base calss for PjRtLoadedExecutable as of https://github.com/tensorflow/tensorflow/commit/a9701644665aa06f8314ea1ce03c0240a7705f50
- Updated parameter type tensorflow::profiler::ToolOptions 
- Added tf_patch for tensorflow::profiler to bypass the profiler related import dependency error

### CODE REVIEW

**Comment by wonjoolee95** [comment]
Currently seeing the error:
```
ERROR: /workspace/pytorch/xla/third_party/tensorflow/tensorflow/compiler/xla/xla_client/BUILD:42:20: Linking tensorflow/compiler/xla/xla_client/libxla_computation_client.so failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/k8-opt/bin/tensorflow/compiler/xla/xla_client/libxla_computation_client.so-2.params
bazel-out/k8-opt/bin/tensorflow/core/profiler/convert/libxplane_to_tools_data.pic.a(xplane_to_tools_data.pic.o): In function `tensorflow::profiler::ConvertMultiXSpacesToToolData(tensorflow::profiler::SessionSnapshot const&, std::basic_string_view<char, std::char_traits<char> >, absl::lts_20220623::flat_hash_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, absl::lts_20220623::container_internal::StringHash, absl::lts_20220623::container_internal::StringEq, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&)':
xplane_to_tools_data.cc:(.text._ZN10tensorflow8profiler29ConvertMultiXSpacesToToolDataERKNS0_15SessionSnapshotESt17basic_string_viewIcSt11char_traitsIcEERKN4absl12lts_2022062313flat_hash_mapINSt7__cxx1112basic_stringIcS6_SaIcEEESt7variantIJiSE_EENS9_18container_internal10StringHashENSH_8StringEqESaISt4pairIKSE_SG_EEEE+0x74): undefined reference to `tensorflow::profiler::ConvertHloProtoToToolData(tensorflow::profiler::SessionSnapshot const&, std::basic_string_view<char, std::char_traits<char> >, absl::lts_20220623::flat_hash_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, absl::lts_20220623::container_internal::StringHash, absl::lts_20220623::container_internal::StringEq, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&)'
collect2: error: ld returned 1 exit status
Target //tensorflow/compiler/xla/xla_client:libxla_computation_client.so failed to build
```
which we suspect to be related to bazel cache. Had no luck with `rm -rf`ing the bazel cache repository. Re-installing in a fresh docker to see if that'll fix this problem.

**Comment by JackCaoG** [comment]
@wonjoolee95 can you rebase this?

**Comment by wonjoolee95** [comment]
> @wonjoolee95 can you rebase this?

Rebased. My local build on a fresh docker is still building TF, currently at:
```
[18,467 / 31,084] 16 actions running
```

**Comment by JackCaoG** [comment]
@wonjoolee95 the fact that it is trying to build `31,084` files suggested you built with xla:GPU. Are you using a gpu docker? It is OK to test it(in fact we need to), but we also want to test the TPUVM build where you build with `XLA_CUDA=0 TPUVM_MODE =1`

**Comment by JackCaoG** [comment]
@ymwangg after this pr merged we should have your latest gpu fix for multi thread compilation. We should make that default for xla:gpu.

**Comment by wonjoolee95** [comment]
> @wonjoolee95 the fact that it is trying to build `31,084` files suggested you built with xla:GPU. Are you using a gpu docker? It is OK to test it(in fact we need to), but we also want to test the TPUVM build where you build with `XLA_CUDA=0 TPUVM_MODE =1`

I see, I'm building locally at my cloudtop using `xla_base:latest` docker. Once the build succeeds, my plan was to test it on TPUVM and gpu docker as well. Should I have just built on TPUVM from the beginning?

**Comment by ymwangg** [comment]
@JackCaoG I'll open a separate PR for that.

**Comment by JackCaoG** [comment]
@wonjoolee95 it is fine really. It is rare that tf build success in one config and fail in other. We need all of them to work before merging anyway.

**Comment by wonjoolee95** [comment]
Hmm, my local build is still ongoing but the CI on this PR also seems to throw the same error:
```
ERROR: /tmp/pytorch/xla/third_party/tensorflow/tensorflow/compiler/xla/xla_client/BUILD:42:20: Linking tensorflow/compiler/xla/xla_client/libxla_computation_client.so failed: (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc @bazel-out/k8-opt/bin/tensorflow/compiler/xla/xla_client/libxla_computation_client.so-2.params
bazel-out/k8-opt/bin/tensorflow/core/profiler/convert/libxplane_to_tools_data.pic.a(xplane_to_tools_data.pic.o): In function `tensorflow::profiler::ConvertMultiXSpacesToToolData(tensorflow::profiler::SessionSnapshot const&, std::basic_string_view<char, std::char_traits<char> >, absl::lts_20220623::flat_hash_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, absl::lts_20220623::container_internal::StringHash, absl::lts_20220623::container_internal::StringEq, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&)':
xplane_to_tools_data.cc:(.text._ZN10tensorflow8profiler29ConvertMultiXSpacesToToolDataERKNS0_15SessionSnapshotESt17basic_string_viewIcSt11char_traitsIcEERKN4absl12lts_2022062313flat_hash_mapINSt7__cxx1112basic_stringIcS6_SaIcEEESt7variantIJiSE_EENS9_18container_internal10StringHashENSH_8StringEqESaISt4pairIKSE_SG_EEEE+0x74): undefined reference to `tensorflow::profiler::ConvertHloProtoToToolData(tensorflow::profiler::SessionSnapshot const&, std::basic_string_view<char, std::char_traits<char> >, absl::lts_20220623::flat_hash_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, absl::lts_20220623::container_internal::StringHash, absl::lts_20220623::container_internal::StringEq, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::variant<int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > > > const&)'
collect2: error: ld returned 1 exit status
```

**Comment by JackCaoG** [comment]
ok time to look up this function and figure out the dependencies 😄

**Comment by JackCaoG** [comment]
`ConvertMultiXSpacesToToolData` in [tensorflow/core/profiler/convert/xplane_to_tools_data.cc](https://github.com/tensorflow/tensorflow/blob/b00661ef3be7aeae01b524b17ecedbc9efbd46b5/tensorflow/core/profiler/convert/xplane_to_tools_data.cc)

calls

`ConvertHloProtoToToolData` in [tensorflow/core/profiler/convert/hlo_to_tools_data.cc](https://github.com/tensorflow/tensorflow/blob/b00661ef3be7aeae01b524b17ecedbc9efbd46b5/tensorflow/core/profiler/convert/hlo_to_tools_data.cc#L84)

not sure why this doesn't work, signature seems to be matching and they are under the same file. Maybe we can start with only build profiler dependency and figure out the issue.

**Comment by JackCaoG** [comment]
Maybe try use bazel to only build https://github.com/tensorflow/tensorflow/blob/b00661ef3be7aeae01b524b17ecedbc9efbd46b5/tensorflow/core/profiler/convert/BUILD#L704 and see if runinto the same error

**Comment by JackCaoG** [comment]
ok nice, seems like tf build part passed, we just need to fix the pt/xla build part.

**Comment by wonjoolee95** [comment]
Yep! Finally, the build passes locally without building CPP tests. Seeing some similar namespace issues I've seen before when I build the CPP tests, working on fixing them now. Should be quick.

**Comment by wonjoolee95** [comment]
Building (including cpp test) succeed locally. Now will need to verify with tests. I'll also open a PR on PyTorch's side with XLA pinned to this branch.

**Comment by JackCaoG** [comment]
@wonjoolee95 can you build on TPUVM(TPUVM_MODE=1) and see if corresponding libtpu works with new tf?

**Comment by wonjoolee95** [comment]
> @wonjoolee95 can you build on TPUVM(TPUVM_MODE=1) and see if corresponding libtpu works with new tf?

Hmm, on a TPUVM building XLA fails while building TF:
```
ERROR: /home/wonjoo/pytorch/xla/third_party/tensorflow/tensorflow/core/tpu/kernels/BUILD:1205:11: Compiling tensorflow/core/tpu/kernels/sharding_util_ops.cc failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 109 arguments skipped)
In file included from ./tensorflow/tsl/platform/default/logging.h:36,
                 from ./tensorflow/core/platform/logging.h:27,
                 from ./tensorflow/core/framework/allocator.h:28,
                 from ./tensorflow/core/framework/op_kernel.h:27,
                 from tensorflow/core/tpu/kernels/sharding_util_ops.cc:26:
./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
   24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
      |                               ^~~~~~~~~~~~~
./tensorflow/core/platform/macros.h:24:31: note: in definition of macro 'TF_ATTRIBUTE_NOINLINE'
   24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
      |                               ^~~~~~~~~~~~~
./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
   24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
      |                               ^~~~~~~~~~~~~
./tensorflow/core/platform/macros.h:24:31: note: in definition of macro 'TF_ATTRIBUTE_NOINLINE'
   24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
      |                               ^~~~~~~~~~~~~
./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
   24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
      |                               ^~~~~~~~~~~~~
./tensorflow/core/platform/macros.h:24:31: note: in definition of macro 'TF_ATTRIBUTE_NOINLINE'
   24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
      |                               ^~~~~~~~~~~~~
Target //tensorflow/compiler/xla/xla_client:libxla_computation_client.so failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 432.890s, Critical Path: 184.68s
INFO: 7279 processes: 711 internal, 6568 local.
FAILED: Build did NOT complete successfully
```

**Comment by wonjoolee95** [comment]
According to https://stackoverflow.com/questions/21116270/gcc-attributes-with-c-methods, seems like an issue due to g++ version. FWIW, my cloudtop (where the build is working) is using `7.5.0` and the TPUVM is using `9.4.0`.

**Comment by JackCaoG** [comment]
I would say let's use docker on TPUVM too. Our wheel is published using a docker build, as long as we can build inside of a docker we are fine..

**Comment by wonjoolee95** [comment]
CI failing with:
```
test_empty_strided (__main__.TestAtenXlaTensor) ... 2022-08-29 21:59:37.855768: F  426514 /tmp/pytorch/xla/third_party/tensorflow/bazel-tensorflow/tensorflow/compiler/xla/shape.h:158] Check failed: has_layout() element_type: TUPLE tuple_shapes { element_type: F32 dimensions: 2 dimensions: 4 dimensions: 1 dimensions: 6 layout { minor_to_major: 3 minor_to_major: 2 minor_to_major: 1 minor_to_major: 0 } is_dynamic_dimension: false is_dynamic_dimension: false is_dynamic_dimension: false is_dynamic_dimension: false } tuple_shapes { element_type: F32 dimensions: 6 dimensions: 2 dimensions: 1 dimensions: 3 layout { minor_to_major: 0 minor_to_major: 1 minor_to_major: 3 minor_to_major: 2 } is_dynamic_dimension: false is_dynamic_dimension: false is_dynamic_dimension: false is_dynamic_dimension: false } tuple_shapes { element_type: F32 dimensions: 6 layout { minor_to_major: 0 } is_dynamic_dimension: false }
*** Received signal 6 ***
...
*** Begin stack trace ***
	tensorflow::CurrentStackTrace[abi:cxx11]()
	
	
	gsignal
	abort
	
	xla::Shape::layout() const
	
	
	
	xla::ShapeUtil::ForEachSubshape(xla::Shape const&, std::function<void (xla::Shape const&, xla::ShapeIndex const&)> const&)
	xla::util::ShapeHash(xla::Shape const&)
	torch_xla::OpByOpExecutor::BuildOps(c10::ArrayRef<torch::lazy::Value>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, absl::lts_20220623::Span<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const>)
	torch_xla::OpByOpExecutor::Execute(c10::ArrayRef<torch::lazy::Value>, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, absl::lts_20220623::Span<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const>)
	
	xla::util::AsyncTask<int>::Schedule()::{lambda()#1}::operator()() const
	
	
	
	clone
*** End stack trace ***
```
Just rebased the PR and the XLA head CIs are passing, so seems like this PR caused this error.

**Comment by wonjoolee95** [comment]
Locally on cloudtop, the failed test runs successfully:
```
(base) jenkins@26d7adccbc26:/workspace/pytorch/xla$ python test/test_operations.py TestAtenXlaTensor.test_empty_strided
/opt/conda/lib/python3.7/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: libc10_cuda.so: cannot open shared object file: No such file or directory
  warn(f"Failed to load image Python extension: {e}")
2022-08-29 22:47:18.495829: W 1175082 tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
2022-08-29 22:47:18.495883: W 1175082 tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:263] failed call to cuInit: UNKNOWN ERROR (303)
----------------------------------------------------------------------
Ran 1 test in 0.821s

OK
```
Going to give a retry.

**Comment by wonjoolee95** [comment]
> > @wonjoolee95 can you build on TPUVM(TPUVM_MODE=1) and see if corresponding libtpu works with new tf?
> 
> Hmm, on a TPUVM building XLA fails while building TF:
> 
> ```
> ERROR: /home/wonjoo/pytorch/xla/third_party/tensorflow/tensorflow/core/tpu/kernels/BUILD:1205:11: Compiling tensorflow/core/tpu/kernels/sharding_util_ops.cc failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 109 arguments skipped)
> In file included from ./tensorflow/tsl/platform/default/logging.h:36,
>                  from ./tensorflow/core/platform/logging.h:27,
>                  from ./tensorflow/core/framework/allocator.h:28,
>                  from ./tensorflow/core/framework/op_kernel.h:27,
>                  from tensorflow/core/tpu/kernels/sharding_util_ops.cc:26:
> ./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
>    24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
>       |                               ^~~~~~~~~~~~~
> ./tensorflow/core/platform/macros.h:24:31: note: in definition of macro 'TF_ATTRIBUTE_NOINLINE'
>    24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
>       |                               ^~~~~~~~~~~~~
> ./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
>    24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
>       |                               ^~~~~~~~~~~~~
> ./tensorflow/core/platform/macros.h:24:31: note: in definition of macro 'TF_ATTRIBUTE_NOINLINE'
>    24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
>       |                               ^~~~~~~~~~~~~
> ./tensorflow/core/platform/macros.h:24:31: error: attributes are not allowed on a function-definition
>    24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
>       |                               ^~~~~~~~~~~~~
> ./tensorflow/core/platform/macros.h:24:31: note: in definition of macro 'TF_ATTRIBUTE_NOINLINE'
>    24 | #define TF_ATTRIBUTE_NOINLINE __attribute__((noinline))
>       |                               ^~~~~~~~~~~~~
> Target //tensorflow/compiler/xla/xla_client:libxla_computation_client.so failed to build
> Use --verbose_failures to see the command lines of failed build steps.
> INFO: Elapsed time: 432.890s, Critical Path: 184.68s
> INFO: 7279 processes: 711 internal, 6568 local.
> FAILED: Build did NOT complete successfully
> ```

Hmm, still seeing this error even on docker on TPUVM. Looking into it.

**Comment by wonjoolee95** [comment]
This is odd, I tried updating the gcc/g++ version in my TPUVM docker to match the working version of my cloudtop (gcc+-7), but the TF build still failed with the same error. So I tried building with the original TF pin in PT/XLA master, and the build succeeds. The weird thing is that this file that causes the error `tensorflow/core/platform/macros.h` hasn't changed between the original pin (https://github.com/tensorflow/tensorflow/blob/57d35fb585de65a34cd6871b8ca3a6501183f98a/tensorflow/core/platform/macros.h) and the new pin (https://github.com/tensorflow/tensorflow/blob/b00661ef3be7aeae01b524b17ecedbc9efbd46b5/tensorflow/core/platform/macros.h).

**Comment by JackCaoG** [comment]
on TPUVM docker, if you don't set `TPUVM_MODE=1` would new pin pass the build?

**Comment by wonjoolee95** [comment]
> on TPUVM docker, if you don't set `TPUVM_MODE=1` would new pin pass the build?

Just tried building without `TPUVM_MODE=1` after cleaning cache, still fails with the new pin. 

And the CI error seems real, it happened again (although all the xla changes in this PR looks not too related at a first glance).

**Comment by wonjoolee95** [comment]
> Locally on cloudtop, the failed test runs successfully:
> 
> ```
> (base) jenkins@26d7adccbc26:/workspace/pytorch/xla$ python test/test_operations.py TestAtenXlaTensor.test_empty_strided
> /opt/conda/lib/python3.7/site-packages/torchvision/io/image.py:13: UserWarning: Failed to load image Python extension: libc10_cuda.so: cannot open shared object file: No such file or directory
>   warn(f"Failed to load image Python extension: {e}")
> 2022-08-29 22:47:18.495829: W 1175082 tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/cuda/lib64:/usr/local/nvidia/lib:/usr/local/nvidia/lib64
> 2022-08-29 22:47:18.495883: W 1175082 tensorflow/compiler/xla/stream_executor/cuda/cuda_driver.cc:263] failed call to cuInit: UNKNOWN ERROR (303)
> ----------------------------------------------------------------------
> Ran 1 test in 0.821s
> 
> OK
> ```
> 
> Going to give a retry.

So it seems like running the test with `TestAtenXlaTensor.test_empty_strided` with `run_dynamic` flag succeeds but running with `run_opbyop` fails. I can reproduce the failure with `run_opbyop` locally on cloudtop.

**Comment by JackCaoG** [comment]
failure seems real, let me know if you need help debugging.

**Comment by JackCaoG** [comment]
Able to repo the failure and it only happens under opbyop mode. It seems like hash function trying to access the layout of a shape but that shape does not have layout
```
54│ torch::lazy::hash_t Hash(const xla::Shape& shape) {
55├>  auto shape_hash = xla::util::ShapeHash(shape);
56│   return c10::uint128(absl::Uint128High64(shape_hash),
57│                       absl::Uint128Low64(shape_hash));
```

```
(gdb) p shape.has_layout()
[New Thread 0x7ff8b1a11700 (LWP 22824)]
$1 = false
```

which trigger the check in 
```
 const Layout& layout() const {
    CHECK(has_layout()) << ShortDebugString();
    return *layout_;
  }
```

**Comment by JackCaoG** [comment]
I think I need to update 
https://github.com/pytorch/xla/blob/08ae1044c2a7e314895f9946104cbe399e096515/third_party/xla_client/xla_util.cc#L21-L29

to handle the xla::shape without layout case.

**Comment by JackCaoG** [comment]
ok in https://github.com/tensorflow/tensorflow/commit/2e9e7bf7801b448fce8985eb08a903a8cef5516d layout becomes optional which cause the issue above.

**Comment by JackCaoG** [comment]
I tried on XRT but then get a different error
``
2022-09-03 00:35:55.678461: E  256910 tensorflow/core/tpu/kernels/tpu_compilation_cache_external.cc:113] LiteralProto has no layout
2022-09-03 00:35:55.678672: F  256910 tensorflow/core/tpu/kernels/tpu_program_group.cc:86] Check failed: xla_tpu_programs.size() > 0 (0 vs. 0)
https://symbolize.stripped_domain/r/?trace=7f5f128b67bb,7f5f12a5172f,7f5ee34ec612,7f5ee989833a,7f5ee98dd4b0,7f5ee98ddee5,7f5ee5934704,7f5eeae6d688,7f5eef27e65f,7f5eef27f5f7,7f5eefa24c75,7f5eefa22b72,7f5eefa04e74,7f5f12a46fa2&map=b16d3ac9b663f9fe3e0b28626c7848a5c80c15c4:7f5ee08d7000-7f5ef3073040 
*** SIGABRT received by PID 256063 (TID 256910) on cpu 16 from PID 256063; stack trace: ***
PC: @     0x7f5f128b67bb  (unknown)  raise
    @     0x7f5edfc50f34        976  (unknown)
    @     0x7f5f12a51730       3904  (unknown)
    @     0x7f5ee34ec613        928  tensorflow::tpu::TpuProgramGroup::Initialize()
    @     0x7f5ee989833b       1728  tensorflow::tpu::TpuCompilationCacheExternal::InitializeEntry()
    @     0x7f5ee98dd4b1       1104  tensorflow::tpu::TpuCompilationCacheInterface::CompileIfKeyAbsentHelper()
    @     0x7f5ee98ddee6        128  tensorflow::tpu::TpuCompilationCacheInterface::CompileIfKeyAbsent()
    @     0x7f5ee5934705       1360  tensorflow::XRTCompileOp::Compute()
    @     0x7f5eeae6d689        464  tensorflow::XlaDevice::Compute()
    @     0x7f5eef27e660       3024  tensorflow::(anonymous namespace)::ExecutorState<>::Process()
    @     0x7f5eef27f5f8         48  std::_Function_handler<>::_M_invoke()
    @     0x7f5eefa24c76        160  Eigen::ThreadPoolTempl<>::WorkerLoop()
    @     0x7f5eefa22b73         64  std::_Function_handler<>::_M_invoke()
    @     0x7f5eefa04e75         80  tensorflow::(anonymous namespace)::PThread::ThreadFn()
    @     0x7f5f12a46fa3  (unknown)  start_thread
https://symbolize.stripped_domain/r/?trace=7f5f128b67bb,7f5edfc50f33,7f5f12a5172f,7f5ee34ec612,7f5ee989833a,7f5ee98dd4b0,7f5ee98ddee5,7f5ee5934704,7f5eeae6d688,7f5eef27e65f,7f5eef27f5f7,7f5eefa24c75,7f5eefa22b72,7f5eefa04e74,7f5f12a46fa2&map=b16d3ac9b663f9fe3e0b28626c7848a5c80c15c4:7f5ee08d7000-7f5ef3073040,77627ed46abea7a1203418c705242500:7f5eca65c000-7f5edffe4ba0 
E0903 00:35:55.932123  256910 coredump_hook.cc:370] RAW: Remote crash data gathering hook invoked.
E0903 00:35:55.932159  256910 client.cc:242] RAW: Coroner client retries enabled (b/136286901), will retry for up to 30 sec.
E0903 00:35:55.932166  256910 coredump_hook.cc:477] RAW: Sending fingerprint to remote end.
E0903 00:35:55.932174  256910 coredump_socket.cc:118] RAW: Stat failed errno=2 on socket /var/google/services/logmanagerd/remote_coredump.socket
E0903 00:35:55.932182  256910 coredump_hook.cc:481] RAW: Cannot send fingerprint to Coroner: [NOT_FOUND] Missing crash reporting socket. Is the listener running?
E0903 00:35:55.932189  256910 coredump_hook.cc:552] RAW: Dumping core locally.
E0903 00:36:05.837435  256910 process_state.cc:774] RAW: Raising signal 6 with default behavior
Aborted (core dumped)`
``

I suspect this has to do with https://github.com/tensorflow/tensorflow/commit/2e9e7bf7801b448fce8985eb08a903a8cef5516d where layout becomes optional..

**Comment by JackCaoG** [comment]
Need to look into how layout should be set by default.

**Comment by JackCaoG** [comment]
libtpu for 08/24 also is not published, so I manually copied it from `gs://cloud-tpu-v2-images-dev-artifacts/libtpu/2022-08-24/libtpu.so`

**Comment by wonjoolee95** [comment]
Huh, this looks good but PyTorch PR (https://github.com/pytorch/pytorch/pull/85071) is failing with:
```
ERROR: /var/lib/jenkins/workspace/xla/third_party/tensorflow/tensorflow/compiler/xla/service/cpu/BUILD:399:11: Compiling tensorflow/compiler/xla/service/cpu/runtime_fp16.cc failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-8/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 32 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from tensorflow/compiler/xla/service/cpu/runtime_fp16.cc:16:
./tensorflow/compiler/xla/service/cpu/runtime_fp16.h:25:23: error: _Float16 is not supported on this target
using XlaF16ABIType = _Float16;
                      ^
```

**Comment by wonjoolee95** [comment]
Searching this error on Google, I can see some related posts like https://discourse.llvm.org/t/how-to-build-compiler-rt-for-new-x86-half-float-abi/63366/4 that suggests adding a new `CXXFLAG`.

**Comment by wonjoolee95** [comment]
Okay, according to https://gcc.gnu.org/onlinedocs/gcc/ARM-Options.html and https://stackoverflow.com/questions/31242106/fp16-type-undefined-in-gnu-arm-c, it looks like the cxx option `-mfp16-format=ieee` may be needed.

**Comment by wonjoolee95** [comment]
Still seeing failures. According to [clang-8 official documentation](https://releases.llvm.org/8.0.0/tools/clang/docs/LanguageExtensions.html), `Float16` is not supported on `x86_64`. And according to our [Dockerfile](https://github.com/pytorch/xla/blob/master/.circleci/docker/Dockerfile#L3), our docker image is `cudnn8-devel-ubuntu18.04` which is `x86_64` noted in https://hub.docker.com/r/nvidia/cuda. So `Float16` error on `x86_64` is known/expected as it's not supported.  

However, it's still a bit odd because PT/XLA CI and upstream PT CI are using the same docker image, but we're only seeing failures in the upstream CI. Looking at the [TF code that's causing this problem](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/cpu/runtime_fp16.h#L23-L25), it seems like it defines `Float16` after some checks for macros, and that might differ between PT/XLA and upstream PT.

**Comment by wonjoolee95** [comment]
> @wonjoolee95 Looking at the error
> 
> ```
> ERROR: /home/jenkins/.cache/bazel/_bazel_jenkins/b463291cb8b07b4bfde1e3a43733cd1a/external/llvm-project/llvm/BUILD.bazel:144:11: Compiling llvm/lib/Demangle/Demangle.cpp failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-8/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 66 arguments skipped)
> 
> Use --sandbox_debug to see verbose messages from the sandbox
> clang: error: unknown argument: '-mfp16-format=ieee'
> ```
> 
> it seems like it does not understand this option

Yeah, the '-mfp16-format=ieee' appears to be `gcc` option only. I saw [https://stackoverflow.com/questions/45108628/how-to-enable-fp16-type-on-gcc-for-x86-64](https://stackoverflow.com/questions/45108628/how-to-enable-fp16-type-on-gcc-for-x86-64) that suggests flags `-fnative-half-type` and  `-fallow-half-arguments-and-returns` for clang. I've added these and waiting on results from the upstream CI.

**Comment by wonjoolee95** [comment]
I kept seeing failures on the PyTorch upstream with all the different compiler flags mentioned above. So I tried creating another [TF patch](https://github.com/pytorch/xla/pull/3922/commits/2b9ea2c98f513023304cbb9c1c4797bf6719d776) that undos the [recent TF commit](https://github.com/tensorflow/tensorflow/commit/d5fd9df089124289f8f294f4cf694561935c97f4) that's included in this pin update. Now, the [upstream CI is passing](https://github.com/pytorch/pytorch/pull/85071) (there's is a unrelated CI failing but the `xla` CIs are passing). 

With this TF patch, I can still confirm that sanity tests and resnet (with `--fake-data`) passes successfully. @JackCaoG, do you think adding this patch is okay?

**Comment by JackCaoG** [comment]
@snadampal We have to revert https://github.com/tensorflow/tensorflow/commit/d5fd9df089124289f8f294f4cf694561935c97f4 in the tf build to get upstream CI build pytorch/xla, is this change necessary for you?

File: torch_xla/csrc/init_python_bindings.cpp
Comment (JackCaoG) [inline torch_xla/csrc/init_python_bindings.cpp:783]: what's this change for?

**Comment by snadampal** [comment]
@JackCaoG , the change looks related to fp16 and may not be required for the acl backend feature I'm interested in. Thanks for checking.

**Comment by wonjoolee95** [comment]
Thanks for the response, @snadampal. 

@JackCaoG, what do you think, shall we try to merge this one in?

**Review by JackCaoG — APPROVED** [review]

**Pull request merged**
